### DISCUSSION TITLE
Nix Packaging questions

### DISCUSSION THREAD

**Discussion by nixith** [pinnacle-comp/pinnacle#212] [Q&A] [answered]
According to the [tracking issue](https://github.com/pinnacle-comp/pinnacle/issues/142) it looks like a nix package derivation needs to be made, and I should hopefully have some time to help out with that soonish, but I was curious if there were any design decisions that had been made around this before I started writing anything.

Additionally, I'm curious about how getting rust configuration might work on a more minimal system, although that might be out of scope of an initial packaging attempt.

**Comment by Ottatop** [comment]
I have very little knowledge of Nix and Nix packaging, most of the flake was written by @LogicalOverflow.

As for design decisions, not *quite* sure if this addresses what you're asking but
- Configs are in separate files from the binary and get installed to `$XDG_DATA_HOME` along with the protobuf definitions
    - This is currently being done in `build.rs` but is in the process of being moved to an external command so it doesn't implicitly run all the time
- The Lua library is a Luarocks rock, so that needs to be installed
- The Rust API is a separate crate, idk how that works with Nix or if it can be in a separate package

**Reply by nixith** [reply]
It should be able to be in a separate package, flakes can self reference for dependencies (i.e. default is `pinnacle`, api is `api`, pinnacle calls `self.${system}.api` as a dependency). The main issue I was thinking of for the rust config was that you need to be able to actually build it at pinnacle's runtime, same for luarocks. My guess for working around that would be something to do with `metaconfig.toml`'s `command` attribute, maybe patching the cargo or luarocks reference to a nix store path, but that's probably better to postpone after the initial package. 


My other question would then be about dealing with external dependencies. [nixpkgs has a standard for doing rust](https://ryantm.github.io/nixpkgs/languages-frameworks/rust/#compiling-rust-applications-with-cargo), but I know some people prefer using a tool like [naersk](https://github.com/nix-community/naersk), which to my knowledge has a few advantages.

**Reply by Ottatop** [reply, marked as answer]
> My other question would then be about dealing with external dependencies. [nixpkgs has a standard for doing rust](https://ryantm.github.io/nixpkgs/languages-frameworks/rust/#compiling-rust-applications-with-cargo), but I know some people prefer using a tool like [naersk](https://github.com/nix-community/naersk), which to my knowledge has a few advantages.

Pick whatever you think is better, idk what the advantages of using naersk are

**Reply by nixith** [reply]
I'm currently running into some issues with the way things are set up, i.e. build.rs [breaking specification](https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script) and modifying files outside of specified directories. Do you mind if I make a tracking issue to catalogue things as I work through them?

**Reply by Ottatop** [reply]
> Do you mind if I make a tracking issue to catalogue things as I work through them?

Go ahead!

> build.rs [breaking specification](https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script)

Yea I had a feeling I was doing some questionable stuff in the build script, good to know that

**Comment by cassandracomar** [comment]
I got it all working just using upstream `buildRustPackage` and nixos/home-manager modules. [see the PR](https://github.com/pinnacle-comp/pinnacle/pull/352)
